R/AllClasses.R

```diff
@@ -0,0 +1,223 @@
+#' colDataColorMap,ExperimentColorMap,character-method
+#' colDataColorMap,ExperimentColorMap,integer-method
+#' rowDataColorMap,ExperimentColorMap,character-method
+#' rowDataColorMap,ExperimentColorMap,integer-method
+#'
+#' @examples
+#'
+#' # Constructor ----
+#'
+#' count_colors <- viridis::viridis(10)
+#' logcounts_colors <- viridis::magma(10)
+#' fpkm_colors <- viridis::inferno(10)
+#' tpm_colors <- viridis::plasma(10)
+#'
+#' qc_colors <- c("forestgreen", "firebrick1")
+#' names(qc_colors) <- c("Y", "N")
+#'
+#' ecm <- new("ExperimentColorMap",
+#'     assays = list(
+#'         counts = count_colors,
+#'         tophat_counts = count_colors,
+#'         cufflinks_fpkm = logcounts_colors,
+#'         cufflinks_fpkm = fpkm_colors,
+#'         rsem_tpm = tpm_colors
+#'     ),
+#'     colData = list(
+#'         passes_qc_checks_s = qc_colors
+#'     )
+#' )
+#'
+#' # Accessor ----
+#'
+#' assayColorMap(ecm, "logcounts") # viridis::magma(10)
+#' assayColorMap(ecm, "undefined") # default: viridis::viridis(10)
+#'
+#' colDataColorMap(ecm, "passes_qc_checks_s")
+#' colDataColorMap(ecm, "undefined")
+#'
+#' rowDataColorMap(ecm, "undefined")
+#'
+ExperimentColorMap <- function(
+  assays = list(), colData = list(), rowData = list(), ...){
+  new("ExperimentColorMap", assays=assays, colData=colData, rowData=rowData, ...)
+}
+
+# .defaultContinuousColorMap ----
+
+# default continuous colormap
+.defaultContinuousColorMap <- viridis::viridis(10)
+
+# Accessors ----
+
+# assayColorMap ----
+
+setGeneric("assayColorMap", function(x, i, ...) standardGeneric("assayColorMap"))
+
+setMethod("assayColorMap", c("ExperimentColorMap", "character"),
+    function(x, i, ...)
+{
+    .assayColorMap(x, i, ...)
+})
+
+setMethod("assayColorMap", c("ExperimentColorMap", "integer"),
+    function(x, i, ...)
+{
+    .assayColorMap(x, i, ...)
+})
+
+.assayColorMap <- function(x, i, ...){
+  res <- tryCatch({
+        x@assays[[i]]
+    }, error=function(err) {
+        .defaultContinuousColorMap
+    })
+    if (is.null(res)){
+      return(.defaultContinuousColorMap)
+    }
+    return(res)
+}
+
+# colDataColorMap ----
+
+setGeneric("colDataColorMap", function(x, i, ...) standardGeneric("colDataColorMap"))
+
+setMethod("colDataColorMap", c("ExperimentColorMap", "character"),
+    function(x, i, ...)
+{
+      .nonAssayColorMap(x, "colData", i)
+})
+
+setMethod("colDataColorMap", c("ExperimentColorMap", "integer"),
+    function(x, i, ...)
+{
+      .nonAssayColorMap(x, "colData", i)
+})
+
+# rowDataColorMap ----
+
+setGeneric("rowDataColorMap", function(x, i, ...) standardGeneric("rowDataColorMap"))
+
+setMethod("rowDataColorMap", c("ExperimentColorMap", "character"),
+    function(x, i, ...)
+{
+      .nonAssayColorMap(x, "rowData", i)
+})
+
+setMethod("rowDataColorMap", c("ExperimentColorMap", "integer"),
+    function(x, i, ...)
+{
+      .nonAssayColorMap(x, "rowData", i)
+})
+
+# As oppose to assays that are always numeric
```

Comment:
Not necessarily, I could have an assay slot full of strings. Or logicals.

Reply:
Fair enough. Good thing is that we can unit test this outside the app

Reply:
Not sure what you mean by unit test; an assay full of strings or logicals is perfectly valid.

Reply:
Not testing the creation of such arrays, but unit testing the `ExperimentColorMap` behaviour in such cases.
As Fede mentioned, the good thing about having refactored code into functions (and classes) outside the app itself means that we can unit test it and get the code coverage up.